source/extensions/quic_listeners/quiche/envoy_quic_utils.cc

```diff
@@ -105,5 +109,37 @@ Http::StreamResetReason quicErrorCodeToEnvoyResetReason(quic::QuicErrorCode erro
   }
 }
 
+Network::ConnectionSocketPtr
+createConnectionSocket(Network::Address::InstanceConstSharedPtr& peer_addr,
+                       Network::Address::InstanceConstSharedPtr& local_addr,
+                       const Network::ConnectionSocket::OptionsSharedPtr& options) {
+  Network::IoHandlePtr io_handle = peer_addr->socket(Network::Address::SocketType::Datagram);
+  auto connection_socket =
+      std::make_unique<Network::ConnectionSocketImpl>(std::move(io_handle), local_addr, peer_addr);
+  connection_socket->addOptions(Network::SocketOptionFactory::buildIpPacketInfoOptions());
+  connection_socket->addOptions(Network::SocketOptionFactory::buildRxQueueOverFlowOptions());
+  if (options != nullptr) {
+    connection_socket->addOptions(options);
+  }
+  if (!Network::Socket::applyOptions(connection_socket->options(), *connection_socket,
```

Comment:
I vaguely feel like this pattern is repeated all over the place. Can you potentially look and see if there is a way to de-dup with a utility function between the listener sockets and here?

Reply:
Which part? applyOptions->bind->applyOptions? or just applyOptions and failure handling?

Reply:
Not sure I would have to analyze, but it seems like applying options and possibly other code is repeated and could be shared somehow? Worth a quick check.

Reply:
applyOptions() at different socket state and failure handling is kinda similar in ListenerImpl and here. But moving them into one place may lead to behavior change in error handling. I think it's more clear to do it in a seperate PR if desired to do so. That saying, the shared code is basically wrapping Socket::applyOptions() and throw upon failure. It doesn't de-dup many lines of code. Is it worth doing?

Reply:
Sounds good let's leave it for now.